contracts/CollectiveGovernance.sol

```diff
@@ -151,62 +135,194 @@ contract CollectiveGovernance is Governance, VoteStrategy {
     /// @notice veto the current measure
     function veto(uint256 _proposalId)
         public
-        requireStrategyVersion(_proposalId)
         requireElectorSupervisor(_proposalId)
         requireVoteOpen(_proposalId)
+        requireNoVeto(_proposalId)
     {
-        _storage._veto(_proposalId);
+        _storage.veto(_proposalId, msg.sender);
     }
 
     // @notice cast an affirmative vote for the measure
-    function voteFor(uint256 _proposalId)
+    function voteFor(uint256 _proposalId) public requireVoteOpen(_proposalId) requireNoVeto(_proposalId) {
+        VoterClass _class = _storage.voterClass(_proposalId);
+        uint256[] memory _shareList = _class.discover(msg.sender);
+        uint256 count = 0;
+        for (uint256 i = 0; i < _shareList.length; i++) {
+            uint256 shareId = _shareList[i];
+            count += _storage.voteForByShare(_proposalId, msg.sender, shareId);
+        }
+        if (count > 0) {
+            emit VoteTally(_proposalId, msg.sender, count);
+        } else {
+            revert("Not voter");
+        }
+    }
+
+    function voteForWithTokenId(uint256 _proposalId, uint256 _tokenId)
         public
-        requireStrategyVersion(_proposalId)
-        requireVoter(_proposalId, msg.sender)
         requireVoteOpen(_proposalId)
+        requireNoVeto(_proposalId)
     {
-        _storage._castVoteFor(_proposalId, msg.sender);
+        uint256 count = _storage.voteForByShare(_proposalId, msg.sender, _tokenId);
+        if (count > 0) {
+            emit VoteTally(_proposalId, msg.sender, count);
+        } else {
+            revert("Not voter");
+        }
+    }
+
+    function voteForWithTokenList(uint256 _proposalId, uint256[] memory _tokenIdList)
+        external
+        requireVoteOpen(_proposalId)
+        requireNoVeto(_proposalId)
+    {
+        uint256 count = 0;
+        for (uint256 i = 0; i < _tokenIdList.length; i++) {
+            count += _storage.voteForByShare(_proposalId, msg.sender, _tokenIdList[i]);
+        }
+        if (count > 0) {
+            emit VoteTally(_proposalId, msg.sender, count);
+        } else {
+            revert("Not voter");
+        }
     }
 
     // @notice undo any previous vote
-    function undoVote(uint256 _proposalId)
+    function undoVote(uint256 _proposalId) public requireVoteOpen(_proposalId) requireNoVeto(_proposalId) {
```

Comment:
Is undoing something we want?

Reply:
We probably want to build for a broad audience with different needs for each community.